File: Core/Model/Impl/Route.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Xml.Serialization;

namespace BinkyRailways.Core.Model.Impl
{
    /// <summary>
    /// Route from one block to another.
    /// </summary>
    public sealed partial class Route : ModuleEntity, IRoute, IActionTriggerSourceInternals
    {
        private readonly Property<EntityRef<Block>> fromBlock;
        private readonly Property<BlockSide> fromBlockSide;
        private readonly Property<EntityRef<Block>> toBlock;
        private readonly Property<BlockSide> toBlockSide;
        private readonly Property<EntityRef<Edge>> fromEdge;
        private readonly Property<EntityRef<Edge>> toEdge;
        private readonly JunctionWithStateSet crossingJunctions;
        private readonly EventSet events;
        private readonly EnteringDestinationSensorSet _enteringSensors;
        private readonly ReachedDestinationSensorSet _reachedSensors;
        private readonly Property<int> speed;
        private readonly Property<int> chooseProbability;
        private LocStandardPredicate permissions;
        private readonly Property<bool> closed;
        private readonly Property<int> maxDuration;
        private readonly ActionTrigger enteringDestinationTrigger;
        private readonly ActionTrigger destinationReachedTrigger;

        /// <summary>
        /// Default ctor
        /// </summary>
        public Route()
        {
            fromBlock = new Property<EntityRef<Block>>(this, null);
            fromBlockSide = new Property<BlockSide>(this, BlockSide.Front);
            toBlock = new Property<EntityRef<Block>>(this, null);
            toBlockSide = new Property<BlockSide>(this, BlockSide.Back);
            fromEdge = new Property<EntityRef<Edge>>(this, null);
            toEdge = new Property<EntityRef<Edge>>(this, null);
            crossingJunctions = new JunctionWithStateSet(this);
            events = new EventSet(this);
            _enteringSensors = new EnteringDestinationSensorSet(this);
            _reachedSensors = new ReachedDestinationSensorSet(this);
            speed = new Property<int>(this, DefaultValues.DefaultRouteSpeed);
            chooseProbability = new Property<int>(this, DefaultValues.DefaultRouteChooseProbability);
            permissions = new LocStandardPredicate();
            permissions.EnsureId();
            closed = new Property<bool>(this, false, null);
            maxDuration = new Property<int>(this, DefaultValues.DefaultRouteMaxDuration);
            enteringDestinationTrigger = new ActionTrigger(this, Strings.TriggerNameEnteringDestination);
            destinationReachedTrigger = new ActionTrigger(this, Strings.TriggerNameDestinationReached);
        }

        /// <summary>
        /// Human readable description
        /// </summary>
        [XmlIgnore]
        public override string Description
        {
            get
            {
                var x = (From != null) ? From.Description : "?";
                var y = (To != null) ? To.Description : "?";
                return x + " -> " + y;
            }
            set { base.Description = value; }
        }

        /// <summary>
        /// Trigger fired when a loc has starts entering the destination of this route.
        /// </summary>
        public ActionTrigger EnteringDestinationTrigger { get { return enteringDestinationTrigger; } }

        /// <summary>
        /// Should <see cref="EnteringDestinationTrigger"/> be serialized?
        /// </summary>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public bool ShouldSerializeEnteringDestinationTrigger()
        {
            return !enteringDestinationTrigger.IsEmpty;
        }

        /// <summary>
        /// Trigger fired when a loc has reached the destination of this route.
        /// </summary>
        public ActionTrigger DestinationReachedTrigger { get { return destinationReachedTrigger; } }

        /// <summary>
        /// Should <see cref="DestinationReachedTrigger"/> be serialized?
        /// </summary>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public bool ShouldSerializeDestinationReachedTrigger()
        {
            return !destinationReachedTrigger.IsEmpty;
        }

        /// <summary>
        /// Trigger fired when a loc has starts entering the destination of this route.
        /// </summary>
        IActionTrigger IRoute.EnteringDestinationTrigger { get { return enteringDestinationTrigger; } }

        /// <summary>
        /// Trigger fired when a loc has reached the destination of this route.
        /// </summary>
        IActionTrigger IRoute.DestinationReachedTrigger { get { return destinationReachedTrigger; } }

        /// <summary>
        /// Gets all triggers of this entity.
        /// </summary>
        IEnumerable<IActionTrigger> IActionTriggerSource.Triggers
        {
            get
            {
                yield return enteringDestinationTrigger;
                yield return destinationReachedTrigger;
            }
        }

        /// <summary>
        /// Gets the containing railway.
        /// </summary>
        IRailway IActionTriggerSourceInternals.Railway { get { return Root; } }

        /// <summary>
        /// Gets the persistent entity that contains this action trigger.
        /// </summary>
        IPersistentEntity IActionTriggerSourceInternals.Container { get { return Module; } }

        /// <summary>
        /// Does this entity generate it's own description?
        /// </summary>
        public override bool HasAutomaticDescription { get { return true; } }

        /// <summary>
        /// Starting point of the route
        /// </summary>
        [XmlIgnore]
        public EndPoint From
        {
            get
            {
                Block block;
                if ((fromBlock.Value != null) && (fromBlock.Value.TryGetItem(out block)))
                    return block;
                Edge edge;
                if ((fromEdge.Value != null) && (fromEdge.Value.TryGetItem(out edge)))
                    return edge;
                return null;
            }
            set
            {
                if (From != value)
                {
                    var block = value as Block;
                    var edge = value as Edge;
                    fromBlock.Value = (block != null) ? new EntityRef<Block>(this, block) : null;
                    fromEdge.Value = (edge != null) ? new EntityRef<Edge>(this, edge) : null;
                }
            }
        }

        /// <summary>
        /// Side of the <see cref="From"/> block at which this route will leave that block.
        /// </summary>
        [DefaultValue(BlockSide.Front)]
        public BlockSide FromBlockSide
        {
            get { return fromBlockSide.Value; }
            set { fromBlockSide.Value = value; }
        }

        /// <summary>
        /// End point of the route
        /// </summary>
        [XmlIgnore]
        public EndPoint To
        {
            get
            {
                Block block;
                if ((toBlock.Value != null) && (toBlock.Value.TryGetItem(out block)))
                    return block;
                Edge edge;
                if ((toEdge.Value != null) && (toEdge.Value.TryGetItem(out edge)))
                    return edge;
                return null;
            }
            set
            {
                if (To != value)
                {
                    var block = value as Block;
                    var edge = value as Edge;
                    toBlock.Value = (block != null) ? new EntityRef<Block>(this, block) : null;
                    toEdge.Value = (edge != null) ? new EntityRef<Edge>(this, edge) : null;
                }
            }
        }

        /// <summary>
        /// Side of the <see cref="To"/> block at which this route will enter that block.
        /// </summary>
        [DefaultValue(BlockSide.Back)]
        public BlockSide ToBlockSide
        {
            get { return toBlockSide.Value; }
            set { toBlockSide.Value = value; }
        }

        /// <summary>
        /// Gets the id of the from block.
        /// Used for serialization only.
        /// </summary>
        [Browsable(false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
        public string FromBlock
        {
            get { return fromBlock.GetId(); }
            set { fromBlock.Value = string.IsNullOrEmpty(value) ? null : new EntityRef<Block>(this, value, LookupBlock); }
        }

        /// <summary>
        /// Gets the id of the to block.
        /// Used for serialization only.
        /// </summary>
        [Browsable(false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
        public string ToBlock
        {
            get { return toBlock.GetId(); }
            set { toBlock.Value = string.IsNullOrEmpty(value) ? null : new EntityRef<Block>(this, value, LookupBlock); }
        }

        /// <summary>
        /// Gets the id of the from edge.
        /// Used for serialization only.
        /// </summary>
        [Browsable(false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
        public string FromEdge
        {
            get { return fromEdge.GetId(); }
            set { fromEdge.Value = string.IsNullOrEmpty(value) ? null : new EntityRef<Edge>(this, value, LookupEdge); }
        }

        /// <summary>
        /// Gets the id of the to edge.
        /// Used for serialization only.
        /// </summary>
        [Browsable(false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
        public string ToEdge
        {
            get { return toEdge.GetId(); }
            set { toEdge.Value = string.IsNullOrEmpty(value) ? null : new EntityRef<Edge>(this, value, LookupEdge); }
        }

        /// <summary>
        /// Set of junctions with their states that are crossed when taking this route.
        /// </summary>
        public JunctionWithStateSet CrossingJunctions { get { return crossingJunctions; } }

        /// <summary>
        /// Control serialization of <see cref="CrossingJunctions"/>.
        /// </summary>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public bool ShouldSerializeCrossingJunctions() { return (crossingJunctions.Count > 0); }

        /// <summary>
        /// Events that control this route
        /// </summary>
        public EventSet Events { get { return events; } }

        /// <summary>
        /// Control serialization of <see cref="Events"/>.
        /// </summary>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public bool ShouldSerializeEvents() { return (events.Count > 0); }

        /// <summary>
        /// Set of sensors which (when active) indicate that a loc is has entered
        /// the To block. The loc will then slow down, but not stop.
        /// OBSOLETE
        /// </summary>        
        public SensorRefSet EnteringDestinationSensors { get { return _enteringSensors; } }

        /// <summary>
        /// Control serialization of <see cref="EnteringDestinationSensors"/>.
        /// </summary>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public bool ShouldSerializeEnteringDestinationSensors() { return (_enteringSensors.Count > 0); }

        /// <summary>
        /// Set of sensors which (when active) indicate that a loc is has completely 
        /// reached the To block. The loc will then stop.
        /// OBSOLETE
        /// </summary>
        public SensorRefSet ReachedDestinationSensors { get { return _reachedSensors; } }

        /// <summary>
        /// Control serialization of <see cref="ReachedDestinationSensors"/>.
        /// </summary>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public bool ShouldSerializeReachedDestinationSensors() { return (_reachedSensors.Count > 0); }

        /// <summary>
        /// Speed of locs when going this route.
        /// This value is a percentage of the maximum / medium speed of the loc.
        /// </summary>
        /// <value>0..100</value>
        [DefaultValue(DefaultValues.DefaultRouteSpeed)]
        public int Speed
        {
            get { return speed.Value; }
            set { speed.Value = Math.Max(0, Math.Min(100, value)); }
        }

        /// <summary>
        /// Probability (in percentage) that a loc will take this route.
        /// When multiple routes are available to choose from the route with the highest probability will have the highest
        /// chance or being chosen.
        /// </summary>
        /// <value>0..100</value>
        [DefaultValue(DefaultValues.DefaultRouteChooseProbability)]
        public int ChooseProbability
        {
            get { return chooseProbability.Value; }
            set { chooseProbability.Value = Math.Max(0, Math.Min(100, value)); }
        }

        /// <summary>
        /// Gets the predicate used to decide which locs are allowed to use this route.
        /// </summary>
        public LocStandardPredicate Permissions
        {
            get { return permissions; }
            [EditorBrowsable(EditorBrowsableState.Never)]
            set
            {
                // Store is used only for serialization.
                if (value == null)
                    throw new ArgumentNullException("permissions");
                value.Module = Module;
                value.EnsureId();
                permissions = value;
            }
        }

        /// <summary>
        /// Should Permissions be serialized?
        /// </summary>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public bool ShouldSerializePermissions()
        {
            return !permissions.IsEmpty;
        }

        /// <summary>
        /// Is this route open for traffic or not?
        /// Setting to true, allows for maintance etc. on this route.
        /// </summary>
        [DefaultValue(false)]
        public bool Closed
        {
            get { return closed.Value; }
            set { closed.Value = value; }
        }

        /// <summary>
        /// Maximum time in seconds that this route should take.
        /// If a loc takes this route and exceeds this duration, a warning is given.
        /// </summary>
        [DefaultValue(DefaultValues.DefaultRouteMaxDuration)]
        public int MaxDuration
        {
            get { return maxDuration.Value; }
            set
            {
                if (value < 2)
                    throw new ArgumentOutOfRangeException("Maximum duration is too short");
                maxDuration.Value = value;
            }
        }

        /// <summary>
        /// Does this route references the given block in one of its properties?
        /// </summary>
        internal bool References(Block block)
        {
            return (From == block) || (To == block);
        }

        /// <summary>
        /// Does this route references the given edge in one of its properties?
        /// </summary>
        internal bool References(Edge edge)
        {
            return (From == edge) || (To == edge);
        }

        /// <summary>
        /// Does this route references the given junction in one of its properties?
        /// </summary>
        internal bool References(Junction junction)
        {
            return crossingJunctions.Contains(junction);
        }

        /// <summary>
        /// Starting point of the route
        /// </summary>
        IEndPoint IRoute.From
        {
            get { return From; }
            set { From = (EndPoint)value; }
        }

        /// <summary>
        /// End point of the route
        /// </summary>
        IEndPoint IRoute.To
        {
            get { return To; }
            set { To = (EndPoint) value; }
        }

        /// <summary>
        /// Set of junctions with their states that are crossed when taking this route.
        /// </summary>
        IJunctionWithStateSet IRoute.CrossingJunctions { get { return crossingJunctions.Set; } }

        /// <summary>
        /// Set of events that change the state of the route and it's running loc.
        /// </summary>
        IRouteEventSet IRoute.Events { get { return events.Set; } }

        /// <summary>
        /// Set of sensors which (when active) indicate that a loc is has entered
        /// the To block. The loc will then slow down, but not stop.
        /// </summary>
        //IRouteSensorSet IRoute.EnteringDestinationSensors { get { return enteringSensors.Set; } }

        /// <summary>
        /// Set of sensors which (when active) indicate that a loc is has completely 
        /// reached the To block. The loc will then stop.
        /// </summary>
        //IRouteSensorSet IRoute.ReachedDestinationSensors { get { return reachedSensors.Set; } }

        /// <summary>
        /// Gets the predicate used to decide which locs are allowed to use this route.
        /// </summary>
        ILocStandardPredicate IRoute.Permissions { get { return permissions; } }

        /// <summary>
        /// Module which contains this entity
        /// </summary>
        internal override Module Module
        {
            get { return base.Module; }
            set
            {
                base.Module = value;
                permissions.Module = value;
                events.LinkItems();
            }
        }

        /// <summary>
        /// Accept a visit by the given visitor
        /// </summary>
        public override TReturn Accept<TReturn, TData>(EntityVisitor<TReturn, TData> visitor, TData data)
        {
            return visitor.Visit(this, data);
        }

        /// <summary>
        /// Validate the integrity of this entity.
        /// </summary>
        public override void Validate(IEntity validationRoot, ValidationResults results)
        {
            base.Validate(validationRoot, results);
            toBlock.Validate(validationRoot, results);
            toEdge.Validate(validationRoot, results);
            fromBlock.Validate(validationRoot, results);
            fromEdge.Validate(validationRoot, results);
            enteringDestinationTrigger.Validate(validationRoot, results);
            destinationReachedTrigger.Validate(validationRoot, results);

            if (From == null)
                results.Warn(this, Strings.WarnNoFromEndPointSpecified);
            if (To == null)
                results.Warn(this, Strings.WarnNoToEndPointSpecified);
            if ((To == From) && (To != null)) 
                results.Warn(this, Strings.WarnToEndPointEqualsFromEndPoint);

            crossingJunctions.Validate(validationRoot, results);
            events.Validate(validationRoot, results);
            _enteringSensors.Validate(validationRoot, results);
            _reachedSensors.Validate(validationRoot, results);

            var enterEventCount = events.Count(evt => evt.Behaviors.Any(x => x.StateBehavior == RouteStateBehavior.Enter));
            var reachedEventCount = events.Count(evt => evt.Behaviors.Any(x => x.StateBehavior == RouteStateBehavior.Reached));

            if ((enterEventCount == 0) && (reachedEventCount == 0) && this.IsInternal())
                results.Warn(this, Strings.WarnNoSensorsSpecified);
            else if ((reachedEventCount == 0) && this.IsToInternal())
                results.Warn(this, Strings.WarnNoReachedDestinationSensorsSpecified);
            else if ((enterEventCount == 0) && this.IsToInternal())
                results.Warn(this, Strings.WarnNoEnteringDestinationSensorsSpecified);
        }

        /// <summary>
        /// If this entity uses the given subject, add a <see cref="UsedByInfo"/> entry to 
        /// the given result.
        /// </summary>
        public override void CollectUsageInfo(IEntity subject, UsedByInfos results)
        {
            base.CollectUsageInfo(subject, results);
            if (subject == From)
                results.UsedBy(this, Strings.UsedByRouteAsFrom);
            if (subject == To)
                results.UsedBy(this, Strings.UsedByRouteAsTo);
            if (crossingJunctions.Contains(subject as IJunction)) 
                results.UsedBy(this, Strings.UsedByRouteAsCrossingJunction);
            var @event = events.FirstOrDefault(x => x.Sensor == subject);
            if (@event != null)
            {
                if (@event.Behaviors.Any(x => x.StateBehavior == RouteStateBehavior.Enter))
                    results.UsedBy(this, Strings.UsedByRouteAsEnteringSensor);
                if (@event.Behaviors.Any(x => x.StateBehavior == RouteStateBehavior.Reached))
                    results.UsedBy(this, Strings.UsedByRouteAsReachedSensor);
                if (@event.Behaviors.Any(x => (x.StateBehavior != RouteStateBehavior.Reached) && (x.StateBehavior != RouteStateBehavior.Enter)))
                    results.UsedBy(this, Strings.UsedByRouteAsOtherSensor);
            }
            permissions.CollectUsageInfo(subject, results);
            enteringDestinationTrigger.CollectUsageInfo(subject, results);
            destinationReachedTrigger.CollectUsageInfo(subject, results);
        }

        /// <summary>
        /// The given entity is removed from the package.
        /// </summary>
        protected override void RemovedFromPackage(IPersistentEntity entity)
        {
            base.RemovedFromPackage(entity);
            crossingJunctions.RemovedFromPackage(entity);
            ((IPackageListener)permissions).RemovedFromPackage(entity);
            enteringDestinationTrigger.RemovedFromPackage(entity);
            destinationReachedTrigger.RemovedFromPackage(entity);
        }

        /// <summary>
        /// Human readable name of this type of entity.
        /// </summary>
        public override string TypeName
        {
            get { return Strings.TypeNameRoute; }
        }

        /// <summary>
        /// Upgrade to the latest format
        /// </summary>
        internal override void Upgrade()
        {
            IRoute route = this;
            if (_enteringSensors.Any())
            {
                foreach (var s in _enteringSensors.Set)
                {
                    var @event = route.Events.Add(s);
                    var b = @event.Behaviors.Add();
                    b.StateBehavior = RouteStateBehavior.Enter;
                    b.SpeedBehavior = LocSpeedBehavior.Default;
                }
                _enteringSensors.RemoveAll(x => true);
            }
            if (_reachedSensors.Any())
            {
                foreach (var s in _reachedSensors.Set)
                {
                    var @event = route.Events.Add(s);
                    var b = @event.Behaviors.Add();
                    b.StateBehavior = RouteStateBehavior.Reached;
                    b.SpeedBehavior = LocSpeedBehavior.Default;
                }
                _reachedSensors.RemoveAll(x => true);
            }
        }

        /// <summary>
        /// Lookup a block by id. 
        /// </summary>
        private Block LookupBlock(string id)
        {
            return Module.Blocks[id];
        }

        /// <summary>
        /// Lookup an edge by id.
        /// </summary>
        private Edge LookupEdge(string id)
        {
            return Module.Edges[id];
        }
    }
}

File: pkg/core/model/impl/route.go
// Copyright 2020 Ewout Prangsma
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author Ewout Prangsma
//

package impl

import (
	"encoding/xml"
	"fmt"

	"github.com/binkyrailways/BinkyRailways/pkg/core/model"
	"github.com/binkyrailways/BinkyRailways/pkg/core/model/refs"
)

type route struct {
	routeFields
}

type routeFields struct {
	moduleEntity
	FromBlock                  BlockRef             `xml:"FromBlock"`
	ToBlock                    BlockRef             `xml:"ToBlock"`
	FromBlockSide              *model.BlockSide     `xml:"FromBlockSide,omitempty"`
	ToBlockSide                *model.BlockSide     `xml:"ToBlockSide,omitempty"`
	Speed                      *int                 `xml:"Speed"`
	ChooseProbability          *int                 `xml:"ChooseProbability"`
	MaxDuration                *int                 `xml:"MaxDuration"`
	Closed                     *bool                `xml:"Closed"`
	Events                     routeEventSet        `xml:"Events"`
	CrossingJunctions          junctionWithStateSet `xml:"CrossingJunctions"`
	Outputs                    outputWithStateSet   `xml:"Outputs"`
	Permissions                locStandardPredicate `xml:"Permissions"`
	EnteringDestinationTrigger actionTrigger        `xml:"EnteringDestinationTrigger"`
	DestinationReachedTrigger  actionTrigger        `xml:"DestinationReachedTrigger"`
}

func (rf *routeFields) SetRoute(r *route) {
	rf.Events.SetContainer(r)
	rf.CrossingJunctions.SetContainer(r)
	rf.Outputs.SetContainer(r)
	rf.Permissions.SetContainer(r)
	rf.EnteringDestinationTrigger.Initialize(r, "Entering destination")
	rf.DestinationReachedTrigger.Initialize(r, "Destination reached")
}

var _ model.Route = &route{}

// newRoute initialize a new route
func newRoute() *route {
	r := &route{}
	r.routeFields.SetRoute(r)
	r.SetDescription("New route")
	return r
}

// Accept a visit by the given visitor
func (r *route) Accept(v model.EntityVisitor) interface{} {
	return v.VisitRoute(r)
}

// UnmarshalXML unmarshals any persistent entity.
func (r *route) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
	if err := d.DecodeElement(&r.routeFields, &start); err != nil {
		return err
	}
	r.routeFields.SetRoute(r)
	return nil
}

// Description of this route
func (r *route) GetDescription() string {
	from, to := "?", "?"
	if x := r.GetFrom(); x != nil {
		from = x.GetDescription()
	}
	if x := r.GetTo(); x != nil {
		to = x.GetDescription()
	}
	return fmt.Sprintf("%s -> %s", from, to)
}

// Starting point of the route
func (r *route) GetFrom() model.EndPoint {
	if result, ok := r.FromBlock.Get(r.GetModule()); ok {
		return result
	}
	// TODO Edge
	return nil
}
func (r *route) SetFrom(value model.EndPoint) error {
	if value == nil {
		r.FromBlock.Set(nil, r.GetModule(), r.OnModified)
	} else if b, ok := value.(model.Block); ok {
		return r.FromBlock.Set(b, r.GetModule(), r.OnModified)
	}
	return nil
}

// Side of the <see cref="From"/> block at which this route will leave that block.
func (r *route) GetFromBlockSide() model.BlockSide {
	return refs.BlockSideValue(r.FromBlockSide, model.BlockSideFront)
}
func (r *route) SetFromBlockSide(value model.BlockSide) error {
	if r.GetFromBlockSide() != value {
		r.FromBlockSide = refs.NewBlockSide(value)
		r.OnModified()
	}
	return nil
}

// End point of the route
func (r *route) GetTo() model.EndPoint {
	if result, ok := r.ToBlock.Get(r.GetModule()); ok {
		return result
	}
	// TODO Edge
	return nil
}
func (r *route) SetTo(value model.EndPoint) error {
	if value == nil {
		r.ToBlock.Set(nil, r.GetModule(), r.OnModified)
	} else if b, ok := value.(model.Block); ok {
		return r.ToBlock.Set(b, r.GetModule(), r.OnModified)
	}
	return nil
}

// Side of the <see cref="To"/> block at which this route will enter that block.
func (r *route) GetToBlockSide() model.BlockSide {
	return refs.BlockSideValue(r.ToBlockSide, model.BlockSideBack)
}
func (r *route) SetToBlockSide(value model.BlockSide) error {
	if r.GetToBlockSide() != value {
		r.ToBlockSide = refs.NewBlockSide(value)
		r.OnModified()
	}
	return nil
}

// Set of junctions with their states that are crossed when taking this route.
func (r *route) GetCrossingJunctions() model.JunctionWithStateSet {
	return &r.CrossingJunctions
}

// Set of outputs with their states that are set when taking this route.
func (r *route) GetOutputs() model.OutputWithStateSet {
	return &r.Outputs
}

// Set of events that change the state of the route and it's running loc.
func (r *route) GetEvents() model.RouteEventSet {
	return &r.Events
}

// Speed of locs when going this route.
// This value is a percentage of the maximum / medium speed of the loc.
// <value>0..100</value>
func (r *route) GetSpeed() int {
	return refs.IntValue(r.Speed, model.DefaultRouteSpeed)
}
func (r *route) SetSpeed(value int) error {
	if r.GetSpeed() != value {
		r.Speed = refs.NewInt(value)
		r.OnModified()
	}
	return nil
}

// Probability (in percentage) that a loc will take this route.
// When multiple routes are available to choose from the route with the highest probability will have the highest
// chance or being chosen.
// <value>0..100</value>
func (r *route) GetChooseProbability() int {
	return refs.IntValue(r.ChooseProbability, model.DefaultRouteChooseProbability)
}
func (r *route) SetChooseProbability(value int) error {
	if r.GetChooseProbability() != value {
		r.ChooseProbability = refs.NewInt(value)
		r.OnModified()
	}
	return nil
}

// Gets the predicate used to decide which locs are allowed to use this route.
func (r *route) GetPermissions() model.LocStandardPredicate {
	return &r.Permissions
}

// Sets the predicate used to decide which locs are allowed to use this route.
func (r *route) SetPermissions(value model.LocStandardPredicate) error {
	r.Permissions.CopyFrom(value)
	return nil
}

// Is this route open for traffic or not?
// Setting to true, allows for maintance etc. on this route.
func (r *route) GetClosed() bool {
	return refs.BoolValue(r.Closed, false)
}
func (r *route) SetClosed(value bool) error {
	if r.GetClosed() != value {
		r.Closed = refs.NewBool(value)
		r.OnModified()
	}
	return nil
}

// Maximum time in seconds that this route should take.
// If a loc takes this route and exceeds this duration, a warning is given.
func (r *route) GetMaxDuration() int {
	return refs.IntValue(r.MaxDuration, model.DefaultRouteMaxDuration)
}
func (r *route) SetMaxDuration(value int) error {
	if r.GetMaxDuration() != value {
		r.MaxDuration = refs.NewInt(value)
		r.OnModified()
	}
	return nil
}

// GetTriggers returns all the triggers of this route
func (r *route) GetTriggers() []model.ActionTrigger {
	return []model.ActionTrigger{r.GetEnteringDestinationTrigger(), r.GetDestinationReachedTrigger()}
}

// Trigger fired when a loc has starts entering the destination of this route.
func (r *route) GetEnteringDestinationTrigger() model.ActionTrigger {
	return &r.EnteringDestinationTrigger
}

// Trigger fired when a loc has reached the destination of this route.
func (r *route) GetDestinationReachedTrigger() model.ActionTrigger {
	return &r.DestinationReachedTrigger
}

// Ensure implementation implements Route
func (*route) ImplementsRoute() {
	// Nothing here
}
